.bazelrc

```diff
@@ -23,24 +23,31 @@ build --flag_alias=gcc_version=//:gcc_version
 # Don't include source level debug info on macOS. See
 # https://github.com/urbit/urbit/issues/5561 and
 # https://github.com/urbit/vere/issues/131.
-build:linux --copt='-g'
+build:linux --per_file_copt='//:.*@-g'
 build:linux --host_copt='-g'
 build --strip=never
 
 # Use -O3 as the default optimization level.
-build --copt='-O3'
+build --per_file_copt='//:.*@-O3'
 build --host_copt='-O3'
 
 # Turn on CPU and memory debug for exec config, which we only use to run the
 # fake ship tests.
 build --host_copt='-DU3_CPU_DEBUG'
 build --host_copt='-DU3_MEMORY_DEBUG'
 
+# Set as per-file copts as a plain --copt gets passed to third party
+# dependencies which forces recompilation (slow) if you are switching between
+# including/excluding a symbol define
+build:mem_dbg --per_file_copt='//:.*@-DU3_MEMORY_DEBUG'
+build:cpu_dbg --per_file_copt='//:.*@-DU3_CPU_DEBUG'
+
 # Enable maximum debug info and disable optimizations for debug config. It's
 # important that these lines come after setting the default debug and
 # optimization level flags above.
-build:dbg --copt='-O0'
-build:dbg --copt='-g3'
+build:dbg --copt="-O0"
+# build:dbg --per_file_copt='//:.*@-O0' # ;;: this breaks the build of softfloat for some reason I haven't investigated yet.
```

Comment:
If you remove line 48 and uncomment line 49, `:urbit` builds without issue. Line 48 passes `-O0` to all targets, not just the ones in `//pkg/...`, and I suspect softfloat has an issue with `-O0`, it being an optimized arithmetic library.

Reply:
Did you pass the `--config=dbg` argument when building? Uncommenting line 49 is what breaks debug builds for me.

```
ERROR: /Users/d/src/urbit/vere/pkg/vere/BUILD.bazel:141:10: Linking pkg/vere/urbit failed: (Exit 1): clang failed: error executing command (from target //pkg/vere:urbit) /usr/bin/clang -o bazel-out/darwin_arm64-fastbuild/bin/pkg/vere/urbit bazel-out/darwin_arm64-fastbuild/bin/pkg/vere/_objs/urbit/main.o ... (remaining 36 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
Undefined symbols for architecture arm64:
  "_softfloat_countLeadingZeros16", referenced from:
      _softfloat_normSubnormalF16Sig in libsoftfloat_aarch64.a(s_normSubnormalF16Sig.o)
      _softfloat_subMagsF16 in libsoftfloat_aarch64.a(s_subMagsF16.o)
  "_softfloat_countLeadingZeros32", referenced from:
      _softfloat_mulAddF16 in libsoftfloat_aarch64.a(s_mulAddF16.o)
      _softfloat_normSubnormalF32Sig in libsoftfloat_aarch64.a(s_normSubnormalF32Sig.o)
      _softfloat_shiftNormSigF128M in libsoftfloat_aarch64.a(s_shiftNormSigF128M.o)
      _softfloat_subMagsF16 in libsoftfloat_aarch64.a(s_subMagsF16.o)
      _softfloat_subMagsF32 in libsoftfloat_aarch64.a(s_subMagsF32.o)
      _softfloat_normRoundPackMToF128M in libsoftfloat_aarch64.a(s_normRoundPackMToF128M.o)
      _softfloat_normRoundPackToF32 in libsoftfloat_aarch64.a(s_normRoundPackToF32.o)
      ...
  "_softfloat_countLeadingZeros64", referenced from:
      _softfloat_mulAddF32 in libsoftfloat_aarch64.a(s_mulAddF32.o)
      _softfloat_mulAddF64 in libsoftfloat_aarch64.a(s_mulAddF64.o)
      _softfloat_normSubnormalF64Sig in libsoftfloat_aarch64.a(s_normSubnormalF64Sig.o)
      _softfloat_subMagsF64 in libsoftfloat_aarch64.a(s_subMagsF64.o)
      _softfloat_normRoundPackToF64 in libsoftfloat_aarch64.a(s_normRoundPackToF64.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Target //pkg/vere:urbit failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 58.815s, Critical Path: 54.50s
INFO: 654 processes: 9 internal, 645 darwin-sandbox.
FAILED: Build did NOT complete successfully

```

Reply:
@mcevoypeter response to above?

Reply:
I see what you're saying. I just ran the following:
```
$ bazel build --config=dbg --subcommands --per_file_copt='//:.*@-O0'  :urbit
/usr/local/x86_64-linux-musl/bin/x86_64-linux-musl-gcc -MD -MF bazel-out/k8-fastbuild/bin/external/softfloat/_objs/softfloat_x86_64/s_propagateNaNF128UI.d <snip> -O2 -g -O3 -g3 -O0 <snip>
```

```console
$ bazel build --config=dbg --subcommands --copt='-O0' :urbit
/usr/local/x86_64-linux-musl/bin/x86_64-linux-musl-gcc -MD -MF bazel-out/k8-fastbuild/bin/external/softfloat/_objs/softfloat_x86_64/extF80M_to_i32_r_minMag.d <snip> -O2 -g -O3 -g3 <snip>
```
Counterintuitively, `--copt='-O0'` doesn't append `-O0` to softfloat's compilation commands, but `--per_file_copt='//:.*@-O0'` does.

Reply:
The only explanation I can think of is that `--copt` doesn't override `copts` attribute of `cc_library`/`cc_binary` targets when it recognizes conflicting options, but `--per_file_copt` does. The documentation doesn't make this explicit, but this is what I'm going off of:

[`--copt` documentation](https://bazel.build/docs/user-manual#copt):
> Also note that copts values listed in specific cc_library or cc_binary build rules will be placed on the compiler command line after these options.

[`--per_file_copt` documentation](https://bazel.build/docs/user-manual#per-file-copt):
> When present, any C++ file with a label or an execution path matching one of the inclusion regex expressions and not matching any of the exclusion expressions will be built with the given options.

Reply:
Looked closer at the compile commands you sent. Yeah I see what you're saying. Any clue why -O3 is getting appended there in both cases? That shouldn't be happening at all. The optimization level we define in softfloat.BUILD is -O2. FYI when you give multiple values of the same options like this, gcc takes the last given, so that binary got compiled at -O3

@mcevoypeter

Reply:
If that's the behavior of --per_file_copt, maybe our regex can be more restrictive? I.e. whitelist the directories under /pkg??

Reply:
> Any clue why -O3 is getting appended there in both cases?

We specify `-O3` as the default a few lines above in `.bazelrc` with `build --per_file_copt='//:.*@-O3'`. So `--per_file_copt` seems to be more aggressive than `--copt`.

> whitelist the directories under /pkg??

Sounds promising. So the regex would be `//pkg:.*` instead of `//:.*`.

Reply:
Just verifying the more restrictive pattern works. Pushed those changes. This looks ready to merge to me. Thanks @mcevoypeter for the help.